### ISSUE TITLE
IP Subject Alternative Name (SAN) is not parsed properly.

### ISSUE THREAD

**Issue by sureshg** [Ephenodrom/Dart-Basic-Utils#8]
I have an X509 cert with IP addresses in the SAN. Ideally, SAN can support the following types of data
 * DNS
 * IP 
 * URI
 * Email
 * RID
  * DirName
  * OtherName

Except for IP address, everything can be converted to string type (i might be wrong here). But for IP address, the `X509Utils.x509CertificateFromPem(res.certificate.pem).subjectAlternativNames` returns a malformed string.  I had to explicitly convert this to a proper IP format by parsing the integer type.

```dart
  var sanIPs = X509Utils.x509CertificateFromPem(res.certificate.pem)
      .subjectAlternativNames
      .map((n) => Uint8List.fromList(n.codeUnits).join("."))
      .toList();
  print('Cert SAN: ${sanIPs}');
```

I would be great if this library can support the IP SAN also.

**Comment by Ephenodrom** [comment]
@sureshg 

Thank you for opening the issue. Unfortunately, I can't reconstruct the problem. I created a self signed certificate with 2 sans. One is a dns name and the other is an IPv4 address. Both of them are parsed correctly.

You can try it with this certificate or check out the Unit test I added for this case.

```
-----BEGIN CERTIFICATE-----
MIIDsTCCApmgAwIBAgIUDmyV5Jod3FGoVStvzghkIs1xZTcwDQYJKoZIhvcNAQEL
BQAwajELMAkGA1UEBhMCREUxCzAJBgNVBAgMAkJZMQ0wCwYDVQQHDARDaXR5MRAw
DgYDVQQKDAdDb21wYW55MRUwEwYDVQQLDAxEZXZlbG9wZW1lbnQxFjAUBgNVBAMM
DWVwaGVub2Ryb20uZGUwHhcNMTkxMTI2MTA1MjMzWhcNMjExMTI1MTA1MjMzWjBq
MQswCQYDVQQGEwJERTELMAkGA1UECAwCQlkxDTALBgNVBAcMBENpdHkxEDAOBgNV
BAoMB0NvbXBhbnkxFTATBgNVBAsMDERldmVsb3BlbWVudDEWMBQGA1UEAwwNZXBo
ZW5vZHJvbS5kZTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALkzNq/p
4YblpAyXOCZAsTgZau2NvmyJ/B4H4SgGC+A4sJCBRuhsiZTSxUHiF4Y1WkXPNIRJ
wxMU0BLCbpLP3QYJpXwQhRtVJB2I4BWybAS2zzV1s6x6XLJsuJUQxZXN4rBdSocN
+D9v4qsfaIIkN7CXnEmTFlzUh33YqL/lmCDZcIQe4HIGa7ElfgzMLEcDh60AsUrr
JSgIDGuAKfxRaZLYY1wpRm3sxkwSSGKYFZK/H34GDC3/PGqyLPxFHIeLRFwIlJH+
r/puvcKlot8dh3SCDgNGuzszbfN351Kdp01ZHVzoMLVMsC93hux/kfsBFPR7NSO2
tG9ClLZ23CdqcIcCAwEAAaNPME0wCwYDVR0PBAQDAgQwMBMGA1UdJQQMMAoGCCsG
AQUFBwMBMCkGA1UdEQQiMCCCEWFwaS5lcGhlbm9kcm9tLmRlggsxOTIuMTY4LjAu
MTANBgkqhkiG9w0BAQsFAAOCAQEALAS85yXzCcdpN2uPQIx9edN4/0sglP+hVEXA
XtWgjq8uYpPmSUdjPYMDQzg+A+zgLQc8FB2YK8bk/yuaDRZaRt8xhjaXQsRhrKf8
pAvj+Cwzy5Sn+L21a1xd+IJHVSh+B0xF8H7D9G2+KwG4GbEkwprDdCh/Ci39/5qg
eRnR3fZYB8vnirc4pJr8tOoZ3aKsdyFyBhozZNQETihW1vrzegs5/rQlLmSUwb3C
TQiUAMlBpzuKM6oCy833Ed+uCtCwGAj0unysv9o7pYv4dKW4REyBGQ0lJltpN3Cg
MrWLubLgMfDzQmOC3oOeNVBDUsQBdFwJwqEEO658U+/OsQYVlw==
-----END CERTIFICATE-----
```

Can you tell me the certificate that caused this error?

**Comment by Ephenodrom** [comment]
No response from the author of the issue.

**Comment by sureshg** [comment]
Could you try this one - https://github.com/sureshg/dart-native-sample/blob/master/lib/certs/cert.pem ?

**Comment by Ephenodrom** [comment]
Hello @sureshg 
I see that the two ip addresses are written as the following bytes in the certificate : 

- 87 04 7F 00 00 01 
- 87 04 00 00 00 00

The 87 is the tag, 04 is the length of the value. You can check this out at <https://lapo.it/asn1js/>

The strange thing is, that i find no information about this tag. Also the official RFC at <https://tools.ietf.org/html/rfc3641>  does not contain information about it.

Nethertheless i made some changes to have a seperate behavior if the tag is 87. The fix is already live with version 2.3.8.

Two more questions : 

- How was this certificate created ?
- Can you create another certificate with ipv6 ?

**Comment by sureshg** [comment]
I used https://certificatetools.com/ to generate a cert. By the way, thanks for the fix. I will check it.
